Update design doc: focus on load mechanism (#1682)
open-telemetry/opentelemetry-dotnet-instrumentation: docs/design.md

### Unsupported scenarios

- **Applications using Ahead-of-Time (AOT) compilation**:
The current implementation relies on the [host startup hook](https://github.com/dotnet/runtime/blob/main/docs/design/features/host-startup-hook.md)

> or the [CLR Profiler APIs](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling)

Reviewer: ```suggestion
or the [CLR Profiler API](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling)
```

> and neither is support when publishing AOT compiled applications.

Reviewer: ```suggestion
and neither is supported when publishing AOT compiled applications.
```

Reviewer 2: Might be worth spelling out AOT in full the first time.

Author: @theletterf a few lines above AOT is spelled out, do you think is still desirable to spell it out again here?

Errors occurring at application runtime are logged and should never crash the application.

## Zero-touch source code instrumentation workflow

Instrumenting a .NET application, without requiring source code changes,
is done in two distinct phases:

 1. Injecting the [OpenTelemetry .NET SDK](https://github.com/open-telemetry/opentelemetry-dotnet#readme)
 into the application.
 2. Adding and enabling instrumentations as the targeted libraries are loaded
 into the application.

### Injecting the OpenTelemetry .NET SDK

For **.NET** applications the [OpenTelemetry .NET SDK](https://github.com/open-telemetry/opentelemetry-dotnet#readme)
is injected using the [host startup hook](https://github.com/dotnet/runtime/blob/main/docs/design/features/host-startup-hook.md).
This allows the OpenTelemetry .NET SDK to be setup before any application code
is executed.
Although the OpenTelemetry .NET SDK is injected into a .NET application
without using a CLR Profiler, the later is still required to enable
bytecode instrumentations (see next section).

The **.NET Framework** doesn't support the host startup hook.
For .NET Framework applications the OpenTelemetry .NET SDK is injected using the

> [CLR Profiler APIs](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/)
> which allow modifying application code during their execution.

Reviewer: ```suggestion
[CLR Profiler API](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/)
which allows modifying application code during their execution.
```

> which allow modifying application code during their execution.

Author: ab6187388

> ### Adding and enabling instrumentations

Reviewer: ```suggestion
### Library instrumentations
```

There are two broad instrumentations types injected into the applications:

> - **Source Instrumentations**: Instrumentations created on top of API hooks

Author: 935dd0418 - searched other case between ** just found 2, except for proper names like "CLR Profiler" and ".NET Framework"

- **Bytecode Instrumentations**: Instrumentations created for libraries
or frameworks that lack proper hooks or callbacks to allow the collection
of observability data. These instrumentations are enabled by modifying
the application [IL code](https://en.wikipedia.org/wiki/Common_Intermediate_Language)

> during runtime using the [CLR Profiler APIs](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/).

Reviewer: ```suggestion
during runtime using the [CLR Profiler API](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/).
```

as well as support code to run and implement bytecode instrumentations.

- [**CLR Profiler DLL**](../src/OpenTelemetry.AutoInstrumentation.Native):
Native component that implements a CLR Profiler. The CLR Profiler is used to
modify the application [intermediate language](https://en.wikipedia.org/wiki/Common_Intermediate_Language)
(IL), including the IL of packages used by the application, to add and collect
observability data.
On the .NET Framework the CLR Profiler DLL is also responsible for injecting
the **Loader** (see above) during the application start up.

![Overview](./images/architecture-overview.png)

### Bootstrapping

While the initial mechanism for bootstrapping the OpenTelemetry .NET SDK differs
between .NET and .NET Framework applications, giving that the [host startup hook](https://github.com/dotnet/runtime/blob/main/docs/design/features/host-startup-hook.md)
is not available on the .NET Framework,
in both cases the initialization is done by creating one instance of the
`OpenTelemetry.AutoInstrumentation.Loader.Startup` class from the Loader assembly.
When creating this instance the static constructor of the type performs
the following actions:

> - Adds a handler to the
> [`AssemblyResolve` event](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.assemblyresolve?view=net-5.0),
> so that it can add any assembly needed by the SDK itself or by any instrumentation.
> - Runs, through reflection, the `Initialization` method from
>   the `OpenTelemetry.AutoInstrumentation.Instrumentation` type
>   from the Managed Profiler assembly.
>   - The `Initialization` code bootstraps the OpenTelemetry .NET SDK,
>     adding configured processors, exporters, and so on,
>     and setting the mechanisms to enable any configured source instrumentations.

Reviewer 2: ```suggestion
1.  Adds a handler to the [`AssemblyResolve` event](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.assemblyresolve?view=net-5.0),
so that it can add any assembly needed by the SDK itself or by any instrumentation.
2. Runs, through reflection, the `Initialization` method from
  the `OpenTelemetry.AutoInstrumentation.Instrumentation` type
  from the Managed Profiler assembly.
  a. The `Initialization` code bootstraps the OpenTelemetry .NET SDK,
    adding configured processors, exporters, and so on,
    and setting the mechanisms to enable any configured source instrumentations.
```

#### .NET bootstrapping

> .NET applications rely on the [host startup hook](https://github.com/dotnet/runtime/blob/main/docs/design/features/host-startup-hook.md)
> being configured to use the `StartupHook.Initialize()` method from
> the `OpenTelemetry.AutoInstrumentation.StartupHook` assembly.
> The `Initialize` method will load the `OpenTelemetry.AutoInstrumentation.Loader`
> assembly and will create the `OpenTelemetry.AutoInstrumentation.Loader.Startup` instance.

Reviewer 2: ```suggestion
.NET applications rely on the [host startup hook](https://github.com/dotnet/runtime/blob/main/docs/design/features/host-startup-hook.md)
being configured to use the `StartupHook.Initialize()` method from
the `OpenTelemetry.AutoInstrumentation.StartupHook` assembly.
The `Initialize` method loads the `OpenTelemetry.AutoInstrumentation.Loader`
assembly and creates the `OpenTelemetry.AutoInstrumentation.Loader.Startup` instance.
```

#### .NET Framework bootstrapping

.NET Framework bootstrapping is performed on the [CorProfiler::JITCompilationStarted](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilercallback-jitcompilationstarted-method)
callback. When that callback happens the CLR Profiler DLL takes
the following actions:

> - If this the first user module in the current AppDomain,
> the profiler injects the IL calling the Loader `Startup` constructor.
> - If the first method observed by JITCompilationStarted is IIS startup code,
>   the profiler invokes
>   `AppDomain.CurrentDomain.SetData("OpenTelemetry_IISPreInitStart", true)`,
>   so that automatic instrumentations can correctly handle IIS startup scenarios.

Reviewer 2: ```suggestion
1. If the module is the first user module in the current AppDomain,
the profiler injects the IL to call the Loader `Startup` constructor.
- If the first method observed by `JITCompilationStarted` is IIS startup code,
  the profiler invokes
  `AppDomain.CurrentDomain.SetData("OpenTelemetry_IISPreInitStart", true)`,
  so that automatic instrumentations can correctly handle IIS startup scenarios.
```

(I'm guessing it's a sequence.)

> ### Enabling and injecting instrumentations

Reviewer: ```suggestion
### Injecting instrumentations
```

#### Source instrumentation

> Source instrumentations are injected into the target process by adding
> handlers to the [`AssemblyLoad` event](https://learn.microsoft.com/en-us/dotnet/api/system.appdomain.assemblyload?view=net-7.0)
> and when the targeted assembly is loaded triggering the source instrumentation
> initialization code.

Reviewer 2: ```suggestion
Source instrumentations are injected into the target process by adding
handlers to the [`AssemblyLoad` event](https://learn.microsoft.com/en-us/dotnet/api/system.appdomain.assemblyload?view=net-7.0)
and when the targeted assembly loads and triggers the source instrumentation
initialization code.
```

#### Bytecode instrumentation

> The bytecode instrumentations, called "call target" in this repo, relies on
> the JIT recompilation capability of the CLR to rewrite the IL for instrumented
> methods. This adds logic at the beginning and end of the instrumented methods
> to invoke instrumentation written in this repo, and wraps the calls with
> try-catch blocks to prevent instrumentation errors from affecting the normal operation
> of the application. This IL rewrite happens in the following steps:

Reviewer 2: ```suggestion
Bytecode instrumentations, also defined as "call target" in this repo, rely on
the JIT recompilation capability of the CLR to rewrite the IL for instrumented
methods. This adds logic at the beginning and end of the instrumented methods
to invoke instrumentation included in this project, and wraps the calls with
try-catch blocks to prevent instrumentation errors from affecting the normal operation
of the application. This IL code rewrite happens in the following steps:
```

Any reason for using two names? Couldn't we just use "bytecode instrumentation" or something instead of "call target"?

Author: This is more for historical reasons: there were different forms of bytecode instrumentation in the past, one was "call site" the other "call target". You are right: it is time to remove "call target" from this type of document.

Author: a1eee3830
